### DISCUSSION TITLE
Testing guidelines

### DISCUSSION THREAD

**Discussion by Kolobamanacas** [element-plus/element-plus#2987] [Q&A]
Is there any guides on testing Element-Plus apps?

Let's say I'm using [Vue 3 Test Utils](https://next.vue-test-utils.vuejs.org/) with [Jest](https://jestjs.io/) and I want to make a simple test that makes sure a value entered in an input is added to a table. Most modern HTML interface testing tutorials recommend using [data-attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) for tests as IDs and classes have different purposes and are subjects to change.

How do I add a data-attribute to a table and more importantly, to a row when using Element's [el-table](https://element-plus.org/#/en-US/component/table)?

I've tried to put data attribute directly to a component like this:

``` html
<el-table data-test="testTableIdentifier"> </el-table>
```

But there is a couple of issues with this approach:
- The attribute is added to a wrapper div, not to a table in real DOM, forcing user to make additional actions in tests to get real table.
- Element-Plus actually adds more than one `<table />` to the DOM when using `<el-table />`, which is again hardens testing table and rows with data.
- And, most importantly, it's not clear which CSS-selector to use to get table rows with actual data.

So, as I mentioned above, the main two questions are:
- Is there a way to put data-attributes directly to DOM elements when using Element-Plus?
- Or, is the any other workaround or guidelines on testing Element-Plus apps?

Many thanks in advance.

**Comment by jw-foss** [comment]
@Kolobamanacas Sorry for the late response, I think you have brought up a good point that we never implemented such guidelines. There are things that I would like to elaborate:

1. For now ElementPlus is constantly changing, the API is almost stable but for specific components e.g. `ElTable` we are intending to refactor it, so the API for some(in a small scope) components could be unstable. We are expecting releasing the stable version by the end of the year, with most of the legacy bug fixings, new features.
2. Based on the reason above, I think we are able to write up a documentation about how to test ElementPlus components.

**Comment by sunecosuri** [comment]
> Is there a way to put data-attributes directly to DOM elements when using Element-Plus?

I am also very interested in this.
Is there a plan to address the data attribute or is there a related issue link?

Currently, the wrote below code will not be placed.
```vue
<el-date-picker
  data-test-id="start-date"
  type="date"
>
</el-date-picker>
```
